src/sound/asio/sound.cpp

```diff
@@ -1218,3 +1235,75 @@ int64_t CSound::Flip64Bits ( const int64_t iIn )
 
     return iOut;
 }
+
+// Windows Native MIDI support
+void CSound::MidiStart()
+{
+    /* Get the number of MIDI In devices in this computer */
+    iMidiDevs = midiInGetNumDevs();
+    if ( iMidiDevs > MAX_MIDI_DEVS )
+    {
+        iMidiDevs = MAX_MIDI_DEVS;
+    }
+
+    // printf("Found %d MIDI device%s\n", iMidiDevs, iMidiDevs == 1 ? "" : "s");
+
+    // open all connected MIDI devices and set the callback function to handle incoming messages
```

Comment:
Maybe we should have some way to specify the MIDI device, rather than just opening them all?  This approach would likely cause me problems - I want Reaper to have exclusive control of my hardware (audio and MIDI) and use its virtual routing capabilities (or I'd use JACK the same way).  We don't do this for audio - i.e. greedily grab every audio device.

Reply:
I was going by what is already done for CoreAudio on Mac:

https://github.com/jamulussoftware/jamulus/blob/bedd7b0458d4669bb6309ed19a0f5858106fa2a4/src/sound/coreaudio-mac/sound.cpp#L72-L79

And I verified on my Mac that it did indeed hear messages from multiple devices (Korg nanoKontrol2 and Behringer UMC404HD with Midi connected to my piano). So I attempted to be consistent with that behaviour.

The problem we have is that at the Windows API level, device numbers are arbitrary, and depend on the order the devices were connected, or if left connected, the specific ports they are connected to and the order Windows finds those ports on startup. So providing some sort of selection parameter in a predictable way is tricky. It might be possible to use `midiInGetDevCaps()` to fetch the product name for each found device, such as `nanoKONTROL2` or `UMC404HD 192k`, and add command-line or even GUI settings to include or exclude devices based on that (no idea what happens if someone has two devices of the same type). But this is way more complexity than I had envisaged for this PR, and I would prefer that kind of nuance was deferred to a follow-up PR. It also requires the user to actually know what identifying string the device uses.

I have no experience with Reaper or ReaRoute, but I would be very interested to hear if my current approach causes problems with your setup in practice. If it does, we can work out how to address them, but I'd be reluctant to add a lot of complexity "just in case".

Reply:
I use MIDI drums and keyboards.  Those go to Reaper.

If MME has exclusively assigned them to Jamulus, Reaper won't get the notes and I won't be able to use Jamulus.

Reply:
Have you experimented with the PR build to see whether MME does assign the MIDI device exclusively to Jamulus or shared?

Reply:
If I open Reaper first, Jamulus gets no MIDI data.